File: javascript/ip_reputation.js

import dotenv from 'dotenv';
import readline from 'readline'

dotenv.config({ path: '../.env' })

const input = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});


export function isValidIP(ipAddr) {
    return ipAddr.match(new RegExp('^((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$', 'g'));
}

export async function fetchJson(ipAddr) {
    const url = `https://www.virustotal.com/api/v3/ip_addresses/${ipAddr}`
    if (!isValidIP(ipAddr)) {
        throw new Error("Enter a valid IP address.")
    }
    const response = await fetch(url, {
        headers: {
            "accept": "application/json", "x-apikey": process.env.VT_KEY
        }
    });
    if (response.ok) {
        let json = await response.json();
        return json;
    }
    else {
        throw new Error(`Error code: ${response.status}, ${response.statusText}`);
    }
}


function checkIpReputation(ipAddr) {
    fetchJson(ipAddr)
        .then(json => {
            console.log(`\nReputation for ${ipAddr} is: ${json.data.attributes?.reputation} \n`);
            console.log(`\nMore info for ${ipAddr}:\n ${JSON.stringify(json.data, null, '\t')}`);
        });
}

if (typeof jasmine === 'undefined') {
    input.question("Enter IP Address to check reputation: ", (ip) => {
        checkIpReputation(ip);
        input.close();
    });
}

File: python/ip_reputation.py
import re, os, json
import requests as req
from dotenv import load_dotenv

load_dotenv()
key = os.getenv("VT_KEY")


def ip_is_valid(ip_addr: str):
    return re.match(r'^((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$', ip_addr) is not None
    
def check_ip_reputation(ip_addr: str):
    if(ip_is_valid(ip_addr)):
        url = f"https://www.virustotal.com/api/v3/ip_addresses/{ip_addr}"
        headers = {"accept": "application/json", "x-apikey": key}
        response = req.get(url, headers=headers)
        if(response.status_code == 200):
           data = response.json()
           return(json.dumps(data['data'], indent=2))
        else:
            return(f"Error code: {response.status_code}")
    else:
        return("Enter a valid IP address.")
        
    
def main():
    ip_input = input("Enter IP Address to check reputation: ")
    response_data = check_ip_reputation(ip_input)
    print(f"\nReputation for {ip_input} is: ", json.loads(response_data)['attributes']['reputation'])
    print(f"\nMore info for {ip_input}: \n", response_data)



if __name__ == "__main__":
    main()